Implement label templates for db objects

Tener opened the pull request:
As an extension to [RFD 151](https://github.com/gravitational/teleport/blob/master/rfd/0151-database-permission-management.md), this PR adds label templating for the database objects, to be used in database object mapping rules. Any of the spec fields can be referenced in the label value. Multiple values are supported.

changelog: Database object labels will now support templates.

marcoandredinis reviewed (commented):
Can we use https://pkg.go.dev/text/template instead?
It might seem overkill, but I would rather rely on stdlib than creating our own mini-templating engine

Tener commented:
> Can we use https://pkg.go.dev/text/template instead? It might seem overkill, but I would rather rely on stdlib than creating our own mini-templating engine

Well, this ship has kind of sailed: I've reused the templating features that were developed and are in use for roles.

greedy52 commented on lib/srv/db/common/permissions/import_rules.go:
```diff
@@ -61,39 +65,121 @@ func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObject
+		// evaluate multi-trait template.
+		out, err := evalMultiTemplate(value, objectSpecToTraits(spec))
+		if err != nil {
+			continue
```
I don't think it's a good idea to silently drop the error. Ideally, the error should be propagated to the caller. Or at least log the error here. And in case it errors, should we delete the original value? Do we need special handling on `NotFound`

Tener replied:
I think deleting the previous value is better approach, as you are not getting a mixture of labels across different rules. It is a tough case, however, and I do wonder if we should just log the error and refuse to import that particular database object. This would be somewhat safer option: we cannot ever grant access to an object that wasn't imported.

greedy52 replied:
> I think deleting the previous value is better approach, as you are not getting a mixture of labels across different rules. It is a tough case, however, and I do wonder if we should just log the error and refuse to import that particular database object. This would be somewhat safer option: we cannot ever grant access to an object that wasn't imported.

Ya it's tough. I think I would still prefer to delete the label and log an error, so in case there is a higher rule that writes the label again, it can heal itself. Yet another hacky alternative is put an error in the label so users can see that in TAG. 

Is it possible to validate the template parts when the import rule is created? If so, we are more confident that it won't fail here.

Tener replied:
Yes, this is good point, I'll add validation to import rule creation. All other matters aside, this is good UX.

We should still handle error cases, just in case, but given the validation above I think the right call is to simply reject the object entirely, with appropriate log.

Tener replied:
Added. This got somewhat messy, as the dependencies went the wrong way, so I ended up moving a lot of code around. I believe it makes more sense now, along with fixing the accidental cyclic dependencies.

greedy52 commented on lib/srv/db/common/permissions/import_rules.go:
```diff
@@ -61,39 +65,121 @@ func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObject
+			continue
+		}
+
+		out, err := services.ApplyValueTraits(part, traits)
```
there is a TODO in `ApplyValueTraits`
https://github.com/gravitational/teleport/blob/c5529cd5312d5b66fbbf17e492dc59da9934af1c/lib/services/role.go#L632-L640
Should we be worried if that might break here someday?

Tener replied:
I'll look into the validation bit.

Tener replied:
I took a deep dive into validation... oh boy. Long story short, this is an actual problem. Unless we are happy with accepting that `{{whatever.object_kind}}` evaluates happily to e.g. `table`, some changes will be needed. I'm trying to force our libraries to behave, but some changes will be needed.

Tener replied:
All right, this took some work but I've now made some independent code using the underlying library, which doesn't have the problem of mixing up namespaces. Added a few tests on top.

greedy52 commented on rfd/0151-database-permission-management.md:
```diff
@@ -43,12 +43,51 @@ expected values for object names are specified as a list, while the matching
+      values:
+        - '*'
+  mappings:
+    - add_labels:
+        database: '{{obj.database}}'
+        kind: '{{obj.object_kind}}'
+        name: '{{obj.name}}'
+        protocol: '{{obj.protocol}}'
+        schema: '{{obj.schema}}'
+        service_name: '{{obj.database_service_name}}'
```
nit: for consistency, maybe use `xxx_name` like in match and scope?
```suggestion
    - add_labels:
        database_name: '{{obj.database}}'
        kind: '{{obj.object_kind}}'
        object_name: '{{obj.name}}'
        protocol: '{{obj.protocol}}'
        schema_name: '{{obj.schema}}'
        database_service_name: '{{obj.database_service_name}}'
```
And instead of an object name, maybe we can also interpolate `table_name`, `view_name`. I don't have a strong opinion on this though.

Tener replied:
> nit: for consistency, maybe use xxx_name like in match and scope?

Sure, that works. This is just a template anyway, so users can modify as needed.

> And instead of an object name, maybe we can interpolate the name as `table_name`, `view_name`. I don't have a strong opinion on this though.

This is bit tricky: to make it general, we would have to support the templates for the key, too:
```yaml
`{{obj.object_kind}}_name`: `{{obj.name}}`
```

I don't think we want to implement such a feature without good rationale, seems too easy to misuse. We could do something similar with individual matching labels per object kind, but I think it is better to keep the template more  general.

greedy52 commented on rfd/0151-database-permission-management.md:
```diff
@@ -78,6 +117,7 @@ spec:
       add_labels:
         env: prod
         product: WidgetMaster3000
+        local_id: '{{obj.object_kind}}-{{obj.schema}}-{{obj.name}}'
```
Is a multi-var template really needed? For labels matching, it feels easier to me to match multiple single-purpose labels than matching a big label like `local-id`.

Tener replied:
I felt this is more natural given how often database objects are referenced by their fully qualified name, but I'm happy to trim that complexity from the PR if you think this isn't needed. We can always add it later, as an extension: `'{{obj.object_kind}}-{{obj.schema}}-{{obj.name}}'` would be invalid otherwise, so adding support would mean making valid previously invalid configurations.

greedy52 replied:
```yaml
match:
  local_id: 'table-sales-cars'
```
vs 
```yaml
match:
  kind: table
  schema_name: sales
  object_name: cars
```
I still think the latter is easier and more flexible.

That said, I am not super against having the compound var support here. You can keep it if you like. I just think it is not necessary, and the implementation would be easier without it.

Tener replied:
I've dropped multi-interpolation, although we can easily get it back if needed.

greedy52 reviewed (commented):
overall impl looks good but left a question on usage of the multi-var template

marcoandredinis commented on lib/srv/db/common/permissions/import_rules_test.go:
```diff
@@ -404,3 +406,193 @@ func Test_databaseObjectScopeMatch(t *testing.T) {
+}
+
+func Test_splitExpressions(t *testing.T) {
+	tests := []struct {
```
Can we, please, add more tests to this?

`{{}}` and space before or after `{{ obj.name}}` `{{obj.name }}`

Tener replied:
I was sure I've added more tests here, sorry for the original omission. In any case this is now removed, as I've exposed some utility functions I'm using for parsing the text instead.

Tener committed:
Implement label templates for db objects

- Expose template parsing utilities.
- Interpolate label templates using `typical` library.
- Validate label templates as part of import rule validation.
- Refactor code to fix cyclic imports.

Tener committed:
Use database_service_name and object_kind for label key in preset

greedy52 commented on lib/srv/db/common/databaseobjectimportrule/import_rules.go:
```diff
@@ -52,46 +57,122 @@ func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObject
+	return trace.Wrap(err)
+}
+
+func evalTemplate(template string, spec *dbobjectv1.DatabaseObjectSpec) (string, error) {
```
What's the difference between this impl of `evalTemplate` and calling `parse.NewTraitsTemplateExpression` + `expr.Interpolate`?

Tener replied:
One problem is that `parse.NewTraitsTemplateExpression` ultimately calls `newTraitsTemplateParser` which uses `typical.WithInvalidNamespaceHack()`. 

The other is that it hardcodes `external` and `internal` as the only two available variable names, instead of `obj` that I wanted to use here (because it is more natural).

Finally, the traits are built around `[]string` results, instead of `string`. Bypassing `TraitsTemplateExpression.Interpolate` allows us to avoid the singleton list awkwardness.

greedy52 replied:
I see. `NewTraitsTemplateExpression` has `Traits` in it so it's fine if it hardcodes `external` and `internal`. 

I am not a big fan of exposing some internals from `parse`. Do you think if we can generalize this implementation a little bit and keep it in `parse`?

Tener replied:
The underlying library here is `typical`, which is also our own code. `parse` sounds general, but is truly about traits, and I think we should avoid adding extra features there. There is also [`lib/expression`](https://github.com/gravitational/teleport.e/tree/master/lib/expression) library in `teleport.e`, but it is also based on `typical`.

The only thing used from `parse` is `parse.SplitExpression`, which is a newly exposed utility function. I'm strongly considering breaking this dependency completely - revert the changes to `parse` and just reimplement this function locally; it isn't complex and this would disentangle this unnecessary dependency.

greedy52 replied:
> The only thing used from parse is parse.SplitExpression, which is a newly exposed utility function. I'm strongly considering breaking this dependency completely - revert the changes to parse and just reimplement this function locally; it isn't complex and this would disentangle this unnecessary dependency.

Agreed.

Tener replied:
Done, PTAL.

Tener committed:
revert changes to `lib/utils/parse/parse.go`

Tener committed:
Reimplement splitExpression locally for import rules

Tener committed:
Update RFD with recent changes and mark it as implemented

greedy52 commented on lib/srv/db/common/databaseobjectimportrule/import_rules.go:
```diff
@@ -0,0 +1,279 @@
+
+// validateTemplate evaluates the template, checking for potential errors.
+func validateTemplate(template string) error {
+	_, err := evalTemplate(template, nil)
```
the `nil` scared me for a second. so this works because the generated functions will return empty `""` when spec is `nil`? may worth leaving a comment or use an empty spec

Tener replied:
This works because `func (e expression) eval(spec *dbobjectv1.DatabaseObjectSpec) (string, error)` is resilient against null `spec`... which is a somewhat fragile property, now that I think about it. I'll pass an empty spec instead.

greedy52 commented on lib/srv/db/postgres/users.go:
```diff
@@ -210,9 +211,12 @@ func (e *Engine) applyPermissions(ctx context.Context, sessionCtx *common.Sessio
 	counts, countMap = permissions.CountObjectKinds(objsTagged)
-	e.Log.WithField("kind_counts", countMap).WithField("total", len(objsFetched)).Infof("Tagged %v database objects (%v).", len(objsTagged), counts)
+	e.Log.WithField("kind_counts", countMap).WithField("total", len(objsFetched)).Infof("Tagged %v database objects (%v), errors: %v.", len(objsTagged), counts, len(errs))
+	for objName, err := range errs {
+		e.Log.WithField("name", objName).WithError(err).Debug("failed to apply label due to template error")
+	}
```
should we just stop and return the error? I think we are usually confident that Apply should succeed. What cases it might fail? maybe newer auth with new templates in import rule?

Tener replied:
We don't have very tight control on the templates here, I imagine there may be some mechanisms with which invalid templates sneak through especially if we add more functionality in the future, like functions. This is why I don't think we should just return error and abort the login process; we can settle for localised failure here, i.e. lack of permissions for a specific object.

Tener commented on lib/srv/db/common/databaseobjectimportrule/import_rules.go:
```diff
@@ -0,0 +1,279 @@
+// For the object to be returned, it must match at least one rule.
+// The modification consists of application of extra labels, per matching mappings.
+// If there are any errors due to invalid label template, the corresponding objects will be dropped.
+func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObjectImportRule, database types.Database, objs []*dbobjectv1.DatabaseObject) ([]*dbobjectv1.DatabaseObject, map[string]error) {
```
There can be as many errors here as the number of objects - hundreds or more. I'm worried that aggregating this will be messy, so I wanted to report each one separately. I agree that this isn't the easiest API to follow; what would you say to pass logger as an argument to ApplyDatabaseObjectImportRules instead? This way we can call it as we go, without having to return the map.

greedy52 replied:
> I agree that this isn't the easiest API to follow; what would you say to pass logger as an argument to ApplyDatabaseObjectImportRules instead?

Sure. Alternatively, change the function to do one object at a time. I think passing the logger is easier.

Tener replied:
Done.

Tener committed:
rename files, add test coverage, use empty spec for validation

Tener committed:
Log template errors directly with debug logs, log total error count as info.

Tener committed:
Update logs

Tener committed:
Merge branch 'master' into tener/rfd-151-object-label-templates

Tener committed:
Merge branch 'master' into tener/rfd-151-object-label-templates

smallinsky reviewed (commented):
The code LGTM thought I have left one question regarding building own mustache template language vs just using build-in go template engine.

Tener commented on lib/srv/db/common/databaseobjectimportrule/apply.go:
```diff
@@ -0,0 +1,281 @@
+		}
+		sb.WriteString(text)
+	}
+
+	return sb.String(), nil
+}
+
+func applyMappingToObject(mapping *dbobjectimportrulev1.DatabaseObjectImportRuleMapping, spec *dbobjectv1.DatabaseObjectSpec, labels map[string]string) (bool, error) {
+	// the matching is applied to the object spec; existing object labels does not matter
+	if !databaseObjectScopeMatch(mapping.GetScope(), spec) {
+		return false, nil
+	}
+	if !databaseObjectImportMatch(mapping.GetMatch(), spec) {
+		return false, nil
+	}
+
+	for key, value := range mapping.AddLabels {
+		out, err := evalTemplate(value, spec)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		labels[key] = out
+	}
+
+	return true, nil
+}
+
+func matchPattern(pattern, value string) bool {
+	re, err := libutils.CompileExpression(pattern)
+	if err != nil {
+		return false
+	}
+	return re.MatchString(value)
+}
+
+func matchAny(patterns []string, value string) bool {
+	return utils.Any(patterns, func(pattern string) bool {
+		return matchPattern(pattern, value)
+	})
+}
```
Well, that would be my preference too, but switching to `{{.name}}` would be inconsistent with other places in Teleport.

smallinsky replied:
I don't think that we are other places where a  mustache template is used in teleport so {{.name} vs {{name}} is just a UI decision. 


@xinding33 @r0mant  
Any thought/preference on following UI ? 

```yaml
mappings:
    - add_labels:
        database: '{{obj.database}}'
        kind: '{{obj.object_kind}}'
        name: '{{obj.name}}'
        protocol: '{{obj.protocol}}'
        schema: '{{obj.schema}}'
        service_name: '{{obj.database_service_name}}'
```    
 **VS**
```yaml
mappings:
    - add_labels:
        database: '{{.database}}'
        kind: '{{.object_kind}}'
        name: '{{.name}}'
        protocol: '{{.protocol}}'
        schema: '{{.schema}}'
        service_name: '{{.database_service_name}}'
```

Tener replied:
Hmm, there are a few? Most notably roles:

```
    db_users: ['{{email.local(external.email)}}']
    db_names: ['{{external.db_names}}']
    db_roles: ['{{external.db_roles}}']
    db_labels:
      'env': '{{regexp.replace(external.env, "^(staging)$", "$1")}}'
```

smallinsky replied:
> Hmm, there are a few? Most notably roles:

My bad. For me it quite weird that if we already have mustache language support in other places for DB flow we are creating the same thing from scratch instead of reusing already existed implementation.

greedy52 replied:
Ya, we had a few discussions on using existing package. @Tener actually started with it.
https://github.com/gravitational/teleport/pull/38630#discussion_r1507874496
https://github.com/gravitational/teleport/pull/38630#discussion_r1512977376

Tener replied:
Yeah, we are reusing as much from existing implementation as possible given the slightly different context and requirements.

smallinsky replied:
What would you say about  making simplest generic  implementation parse logic  and moving it into common package utils/parse  where the logic can be refused and it quite readable  ?

```go
func evalTemplate(value string, obj any) (string, error) {
	match := reVariable.FindStringSubmatch(value)
	if len(match) == 0 {
		if strings.Contains(value, "{{") || strings.Contains(value, "}}") {
			return "", trace.BadParameter(
				"%q is using template brackets '{{' or '}}', however expression does not parse, make sure the format is {{expression}}",
				value,
			)
		}
		return value, nil
	}
	prefix, value, suffix := strings.TrimLeftFunc(match[1], unicode.IsSpace), match[2], strings.TrimRightFunc(match[3], unicode.IsSpace)
	p, err := predicate.NewParser(predicate.Def{
		GetIdentifier: func(fields []string) (interface{}, error) {
			return predicate.GetFieldByTag(obj, teleport.JSON, fields)
		},
	})

	parsed, err := p.Parse(value)
	if err != nil {
		return "", trace.Wrap(err)
	}
	result, ok := parsed.(string)
	if !ok {
		return "", trace.BadParameter("expected string, got %T", parsed)
	}

	var sb strings.Builder
	sb.WriteString(prefix)
	sb.WriteString(result)
	sb.WriteString(suffix)
	return sb.String(), nil
}
```

Tener replied:
I think having a specific object to reference (here `obj`) is important for extensibility POV. That said, a minor adjustment in `EvalTemplate` restores that:

```golang
func EvalTemplate(value string, namespace map[string]any) (string, error) {
	match := reVariable.FindStringSubmatch(value)
	if len(match) == 0 {
		if strings.Contains(value, "{{") || strings.Contains(value, "}}") {
			return "", trace.BadParameter(
				"%q is using template brackets '{{' or '}}', however expression does not parse, make sure the format is {{expression}}",
				value,
			)
		}
		return value, nil
	}
	prefix, value, suffix := strings.TrimLeftFunc(match[1], unicode.IsSpace), match[2], strings.TrimRightFunc(match[3], unicode.IsSpace)
	p, err := predicate.NewParser(predicate.Def{
		GetIdentifier: func(fields []string) (interface{}, error) {
			obj, found := namespace[fields[0]]
			if !found {
				return nil, trace.NotFound("identifier %q is not defined", strings.Join(fields, "."))
			}
			return predicate.GetFieldByTag(obj, teleport.JSON, fields[1:])
		},
	})
       // ...
}
```

This is neat. 

I'm not convinced, however, that `EvalTemplate` should be added to `parse`. I've expressed my concerns previously (https://github.com/gravitational/teleport/pull/38630#discussion_r1513163706):

> The underlying library here is `typical`, which is also our own code. `parse` sounds general, but is truly about traits, and I think we should avoid adding extra features there. There is also [`lib/expression`](https://github.com/gravitational/teleport.e/tree/master/lib/expression) library in `teleport.e`, but it is also based on `typical`.
> 
> The only thing used from `parse` is `parse.SplitExpression`, which is a newly exposed utility function. I'm strongly considering breaking this dependency completely - revert the changes to `parse` and just reimplement this function locally; it isn't complex and this would remove this unnecessary dependency.

All in all, I just don't think adding a new utility function to `parse`, unused by anything related to traits, is the right call. Thoughts?

Tener replied:
> Personally I would move it to parse because we have already the reVariable defined in this packet and if first directory that I would check to see if I can reuse something before writing a custom parser :-)

> Nevertheless minimal version of EvalTemplate function implementation is small and if we descide to keep the obj reference we should keep it in db package.

I'm not super happy about copying the regexp, but creating this dependency on `parse` doesn't feel good either. Right now, I believe that avoiding the dependency is better. This choice is also supported by various other uses of `predicate.NewParser` and `typical.NewParser`: the overall design choice seems to be to do the parsing independently on relevant places.

> I thought that this was a side effect decision to support `{{obj.object_kind}}` instead of `{{object_kind}}`

I haven't considered using `{{object_kind}}` at all, as it goes against the extensibility goal, but I'm not sure if `{{object_kind}}` would work any better; `internal` and `external` are hard-coded in relevant places.
